### PR TITLE
use custom meta path finder to import 3rd library

### LINKED ISSUE #378: Enhance the way for importing python third-party library
In #364, it created a module called `pylibmgr` to manage and import python third-party library from third-party library root `thirdparty`.

However, the way for import third-party library is appending to `sys.path`, which deteriorate maintainability.
Reference: https://github.com/solvcon/modmesh/pull/364#discussion_r1650208176

modmesh needs a better way to import third-party library!

### PR DESCRIPTION
**Pull request by j8xixo12**
Using custom meta path finder instead of modifying `sys.path` to import 3rd python library into modmesh.

Link to #378

### CODE REVIEW

File: modmesh/view.py [hunk 1]
```diff
@@ -62,15 +60,6 @@ def _load():
...
             globals()[name] = getattr(_vimpl, name)
-    if core.HAS_VIEW:
-        # The viewer is using PUI as a third-party library,
-        # registering it into toggle system
-        tg = core.Toggle.instance
-        tg.add_subkey("viewer")
-        tv = tg.viewer
-        tv.set_string("library", "PUI")
-
-        pylibmgr.load_library(tv.library)
```
Comment (j8xixo12) [inline modmesh/view.py:63]: Since using custom meta path finder, therefore remove this code.

File: modmesh/pylibmgr.py [hunk 2]
```diff
@@ -27,10 +27,35 @@
...
+            _ = os.path.abspath(lib_path[lib_name])
+            pkg_path = os.path.join(_, lib_name)
+            init_path = os.path.join(pkg_path, '__init__.py')
+
+            if not os.path.exists(init_path):
+                return None
+
+            loader = importlib.machinery.SourceFileLoader(lib_name, init_path)
+            spec = importlib.machinery.ModuleSpec(
+                                                  lib_name,
```
Comment (yungyuc) [inline modmesh/pylibmgr.py]: The indentation makes too many lines.  Could you please change it to be more compact:

```python
spec = importlib.machinery.ModuleSpec(
        lib_name, loader, origin=init_path, is_package=True)
```

File: modmesh/pylibmgr.py [hunk 3]
```diff
@@ -70,20 +95,10 @@ def search_library_root(curr_path, lib_root_name, timeout=1.0):
...
-        if lib_name in lib_path:
-            sys.path.append(lib_path[lib_name])
-        else:
-            raise ImportError
-    except ImportError:
-        sys.stderr.write(f'Can not find {lib_name} in library root.\n')
-        sys.exit(0)
 
+_register_mm_path_finder()
+del _register_mm_path_finder
```
Comment (yungyuc) [inline modmesh/pylibmgr.py]: Good to delete the symbol.  Thanks!

File: modmesh/pylibmgr.py [hunk 4]
```diff
@@ -27,10 +27,35 @@
 import sys
 import os
 import time
+import importlib.abc
+import importlib.machinery
 
 lib_path = {}
 
 
+class mm_path_finder(importlib.abc.MetaPathFinder):
```
Comment (yungyuc) [inline modmesh/pylibmgr.py]: Why this class is not using `CamelCase`?

**Review by yungyuc — COMMENTED** [review]
@j8xixo12 I don't understand how do you make `thirdparty` to be added in the module search path.  Could you please describe the mechanism in code comments and/or review comments?

File: modmesh/pylibmgr.py (hunk 2 above)
Comment (j8xixo12) [reply modmesh/pylibmgr.py]: Done !

File: modmesh/pylibmgr.py (hunk 4 above)
Comment (j8xixo12) [reply modmesh/pylibmgr.py]: I have changed it at the latest commit.

File: modmesh/pylibmgr.py [hunk 5]
```diff
@@ -33,7 +33,7 @@
 lib_path = {}
 
 
-class mm_path_finder(importlib.abc.MetaPathFinder):
+class MMPathFinder(importlib.abc.MetaPathFinder):
```
Comment (tigercosmos) [inline modmesh/pylibmgr.py]: Do we need `MM` here?

Comment (yungyuc) [reply modmesh/pylibmgr.py]: Something indicating the meta path class is for modmesh is very helpful in debugging.  A maintainer can inspect the name of the meta path object type in the `sys.meta_path` list.

Comment (j8xixo12) [reply modmesh/pylibmgr.py]: I have changed it from `MM` to `Modmesh`.
`MM` looks weird in camel case style.

Comment (tigercosmos) [reply modmesh/pylibmgr.py]: `Modmesh` looks better.

File: modmesh/pylibmgr.py [hunk 6]
```diff
@@ -27,10 +27,48 @@
...
 import os
 import time
+import importlib.abc
+import importlib.machinery
 
 lib_path = {}
 
 
+class ModmeshPathFinder(importlib.abc.MetaPathFinder):
+    def find_spec(self, lib_name, path, target=None):
```
Comment (yungyuc) [inline modmesh/pylibmgr.py:38]: @j8xixo12 I still don't get it.  How does this `ModmeshPathFinder` locate the directory `thirdparty/`?  It's not written anywhere.  I must miss something and please help me understand.

**Review by yungyuc — COMMENTED** [review]
I still don't know how `thirdparty/` is given to the meta path.

File: modmesh/pylibmgr.py [hunk 7]
```diff
@@ -70,20 +108,10 @@ def search_library_root(curr_path, lib_root_name, timeout=1.0):
             lib_path[item] = os.path.join(_path, item)
 
 
-def load_library(lib_name):
-    """
-    Append library path into python path by user input library name
+def _register_mm_path_finder():
```
Comment (yungyuc) [inline modmesh/pylibmgr.py]: What about we combine the register function with the existing function `search_library_root()` above?  [`modmesh/core.py#L116`](https://github.com/solvcon/modmesh/blob/7e04e975bc71e137cffd626b865c436f43c0abd3/modmesh/core.py#L116) calls `search_library_root()` already.  In this way, we can remove the global variable `lib_path` too (`sys.meta_path` keeps the meta path created in the combined `search_library_root()`.

Is there a drawback in my proposal?

**Review by yungyuc — CHANGES_REQUESTED** [review]
> I still don't know how `thirdparty/` is given to the meta path.

Oh, I got it.  It's in [`modmesh/core.py#L116`](https://github.com/solvcon/modmesh/blob/7e04e975bc71e137cffd626b865c436f43c0abd3/modmesh/core.py#L116):

```python
    # Walk through the thirdparty folder and register all library
    # into a dictionary.
    pylibmgr.search_library_root(os.getcwd(), 'thirdparty')
```

----

- [x] Merge the meta path addition in the existing function `search_library_root()`.

File: modmesh/pylibmgr.py (hunk 7 above)
Comment (j8xixo12) [reply modmesh/pylibmgr.py]: Do you mean that the step of meta path finder registering `sys.meta_path.append(ModmeshPathFinder())` can be put into `search_library_root()` and `lib_path` can be kept in `ModemshPathFinder()` ?

Comment (yungyuc) [reply modmesh/pylibmgr.py]: Yes.

Comment (j8xixo12) [reply modmesh/pylibmgr.py]: Got it, then the proposal looks to me !
Thanks for your suggestion.

**Review by yungyuc — APPROVED** [review]
Thanks. Having the `lib_paths` in `ModmeshMetaPathFinder` makes it much easier to understand the code.

**Pull request merged**